src/utils/hdf-serialize.c

```diff
@@ -0,0 +1,1033 @@
+	if (attr_data == NULL) {
+		fprintf(stderr, "failed to allocate user attributes");
+		rc = 1;
+		goto out;
+	}
+
+	/* Set the data for all attributes */
+	for (i = 0; i < num_attrs; i++) {
+		attr_data[i].attr_name = names[i];
+		attr_data[i].attr_val.p = buffers[i];
+		attr_data[i].attr_val.len = sizes[i];
+	}
+
+	status = H5Dwrite(dset, *usr_attr_memtype, H5S_ALL,
+			  H5S_ALL, H5P_DEFAULT, attr_data);
+	if (status < 0) {
+		fprintf(stderr, "failed to write user attr dset");
+		rc = 1;
+		goto out;
+	}
+
+out:
+	free(attr_data);
+	H5Dclose(dset);
+	H5Tclose(*usr_attr_memtype);
+	H5Sclose(dspace);
+out_no_attrs:
+	return rc;
+}
+
+int
+serialize_daos_metadata(char *filename, daos_prop_t *props,
+			daos_handle_t coh, int num_attrs,
+			char **names, char **buffers, size_t *sizes)
+{
+	int	rc = 0;
+	hid_t	status;
+	hid_t	file_id;
+	hid_t	usr_attr_memtype;
+	hid_t	usr_attr_name_vtype;
+	hid_t	usr_attr_val_vtype;
+
+	fprintf(stderr, "Writing metadata file: %s\n", filename);
+
+	file_id = H5Fcreate(filename, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+	if (file_id < 0) {
+		rc = 1;
+		fprintf(stderr, "failed to create hdf5 file");
+		goto out;
+	}
+	rc = cont_serialize_props(&file_id, props, coh);
+	if (rc != 0) {
+		fprintf(stderr, "failed to serialize cont layout: "DF_RC,
+			DP_RC(rc));
+		goto out;
+	}
+
+	/* serialize usr_attrs if there are any */
+	if (num_attrs > 0) {
+		/* create User Attributes Dataset in daos_metadata file */
+		usr_attr_memtype = H5Tcreate(H5T_COMPOUND, sizeof(usr_attr_t));
+		if (usr_attr_memtype < 0) {
+			rc = 1;
+			fprintf(stderr, "failed to create user attr memtype");
+			goto out;
+		}
+
+		usr_attr_name_vtype = H5Tcopy(H5T_C_S1);
+		if (usr_attr_name_vtype < 0) {
+			rc = 1;
+			fprintf(stderr, "failed to create user attr name "
+				"vtype");
+			goto out;
+		}
+		status = H5Tset_size(usr_attr_name_vtype, H5T_VARIABLE);
+		if (status < 0) {
+			rc = 1;
+			fprintf(stderr, "failed to create user attr name "
+				"vtype");
+			goto out;
+		}
+		usr_attr_val_vtype = H5Tvlen_create(H5T_NATIVE_OPAQUE);
+		if (usr_attr_val_vtype < 0) {
+			rc = 1;
+			fprintf(stderr, "failed to create user attr val vtype");
+			goto out;
+		}
+		status = H5Tinsert(usr_attr_memtype, "Attribute Name",
+				   HOFFSET(usr_attr_t, attr_name),
+				   usr_attr_name_vtype);
+		if (status < 0) {
+			rc = 1;
+			fprintf(stderr, "failed to insert user attr name");
+			goto out;
+		}
+		status = H5Tinsert(usr_attr_memtype, "Attribute Value",
+				   HOFFSET(usr_attr_t, attr_val),
+				   usr_attr_val_vtype);
+		if (status < 0) {
+			rc = 1;
+			fprintf(stderr, "failed to insert user attr val");
+			goto out;
+		}
+		rc = cont_serialize_usr_attrs(&file_id, &usr_attr_memtype,
+					      num_attrs, names, buffers,
+					      sizes, coh);
+		if (rc != 0) {
+			fprintf(stderr, "failed to serialize usr attributes: "
+				""DF_RC, DP_RC(rc));
+			goto out;
+		}
+
+		status = H5Tclose(usr_attr_name_vtype);
+ 		if (status < 0) {
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs